pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,314 @@
+	}
+
+	openstackClientFactory, err := a.openstackClientFactory.NewFactory(credentials)
+	if err != nil {
+		return fmt.Errorf("could not create Openstack client factory: %w", err)
+	}
+
+	securityGroup, err := ensureSecurityGroup(openstackClientFactory, opt)
+	if err != nil {
+		return err
+	}
+
+	err = ensureSecurityGroupRules(openstackClientFactory, opt, securityGroup.ID)
+	if err != nil {
+		return err
+	}
+
+	instance, err := ensureComputeInstance(logger, openstackClientFactory, opt)
+	if err != nil || instance == nil {
+		return err
+	}
+
+	fipid, err := ensurePublicIPAddress(opt, openstackClientFactory)
+	if err != nil {
+		return err
+	}
+
+	ready, err := ensureAssociateFIPWithInstance(openstackClientFactory, instance, fipid)
+	if err != nil || !ready {
+		return err
+	}
+
+	// refresh instance after public ip attached/created
+	instancenstatus, err := getBastionInstance(openstackClientFactory, opt.BastionInstanceName)
+	if err != nil {
+		return err
+	}
+
+	// check if the instance already exists and has an IP
+	endpoints, err := getInstanceEndpoints(&instancenstatus[0], opt)
+	if err != nil {
+		return err
+	}
+
+	if !endpoints.Ready() {
+		return &ctrlerror.RequeueAfterError{
+			// requeue rather soon, so that the user (most likely gardenctl eventually)
+			// doesn't have to wait too long for the public endpoint to become available
+			RequeueAfter: 5 * time.Second,
+			Cause:        fmt.Errorf("bastion instance has no public/private endpoints yet"),
+		}
+	}
+
+	// once a public endpoint is available, publish the endpoint on the
+	// Bastion resource to notify upstream about the ready instance
+	return controller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.client, bastion, func() error {
+		bastion.Status.Ingress = *endpoints.public
+		return nil
+	})
+}
+
+func ensurePublicIPAddress(opt *Options, openstackClientFactory openstaclClient.Factory) (*floatingips.FloatingIP, error) {
+	info, err := getFipbyName(openstackClientFactory, opt.BastionInstanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	if info != nil && info[0].Status == "ACTIVE" {
+		return &info[0], nil
+	}
+
+	logger.Info("creating new bastion Public IP")
+
+	externalFipInfo, err := getExternalNetworkInfoByName(openstackClientFactory, opt.FloatingPoolName)
+	if err != nil {
+		return nil, err
+	}
+
+	createOpts := floatingips.CreateOpts{
+		FloatingNetworkID: externalFipInfo[0].ID,
+		Description:       opt.BastionInstanceName,
+	}
+
+	fip, err := createFloatingIP(openstackClientFactory, createOpts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get (create) public ip address: %w", err)
+	}
+
+	return fip, nil
+}
+
+func ensureComputeInstance(logger logr.Logger, openstackClientFactory openstaclClient.Factory, opt *Options) (*servers.Server, error) {
+	getInstance, err := getBastionInstance(openstackClientFactory, opt.BastionInstanceName)
+	if err != nil || getInstance != nil {
+		return &getInstance[0], err
```

Comment:
Same as the other comments. You need to check if your lists have members at the very least to avoid segfaulting.

Reply:
committed in 508e7fd9